components/batch-system/src/batch.rs

```diff
@@ -256,9 +260,14 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
         }
 
         if batch.is_empty() {
-            self.handler.pause();
-            if let Ok(fsm) = self.fsm_receiver.recv() {
-                return batch.push(fsm);
+            loop {
+                if self.handler.pause() {
+                    if let Ok(fsm) = self.fsm_receiver.recv() {
+                        return batch.push(fsm);
+                    }
+                } else if let Ok(fsm) = self.fsm_receiver.recv_timeout(Duration::from_micros(100)) {
```

Comment:
This looks like magic. Why should it wait for 100 micro seconds if handler is unable to pause? I think this is more related to raftstore instead of batch system itself.

Reply:
Because some work has not finished yet. Maybe the batch system needs a more general function to meet this need?

Reply:
If work has not finished, it should handle it immediately, no reason to wait.

Reply:
This (need) comes from a special type of task:
1. It needs time to finish.
2. It doesn't need (a full on core of) CPU resource to run the job.

In here, the special task is `waiting for deadline, then do fsync`.

If it's a common task, we will spend the whole thread to do that job, but since the `waiting for deadline` part no need to use CPU resource, so we put the thread into sleep (by wait_with_timeout) to reserve CPU resource as an optimization, then allow the thread wake up periodly to continue its job.

I think this need and the code here is genetic, not raftstore-specified.